alibi_detect/saving/validators.py

```diff
@@ -0,0 +1,84 @@
+import sys
+from typing import Any, Generic, Optional, Type, TypeVar, Union, List
+import numpy as np
+from numpy.lib import NumpyVersion
+from pydantic.fields import ModelField
+
+from alibi_detect.utils.frameworks import has_tensorflow, has_pytorch, has_keops, Framework
+
+if has_tensorflow:
+    import tensorflow as tf
+if has_pytorch:
+    import torch
+
+
+def coerce_int2list(value: int) -> List[int]:
+    """Validator to coerce int to list (pydantic doesn't do this by default)."""
+    if not isinstance(value, list):
+        return [value]
+    else:
+        return value
+
+
+# Framework validator (validates `flavour` and `backend` fields)
+def validate_framework(framework: str, field: ModelField) -> str:
+    if (framework == Framework.TENSORFLOW and has_tensorflow) or (framework == Framework.PYTORCH and has_pytorch) or \
+            (framework == Framework.KEOPS and has_keops):
+        return framework
+    elif framework == Framework.SKLEARN:  # sklearn is a core dep
+        return framework
+    else:
+        raise ImportError(f"`{field.name} = '{framework}'` not possible since {framework} is not installed.")
+
+
+# NumPy NDArray pydantic validator type
+# The code below is adapted from https://github.com/cheind/pydantic-numpy.
+T = TypeVar("T", bound=np.generic)
+if NumpyVersion(np.__version__) < "1.22.0" or sys.version_info < (3, 9):
+    class NDArray(Generic[T], np.ndarray):
+        """
+        A Generic pydantic model to coerce to np.ndarray's.
+        """
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, val: Any, field: ModelField) -> np.ndarray:
+            return _coerce2NDArray(cls, val, field)
+
+else:
+    class NDArray(Generic[T], np.ndarray[Any, T]):  # type: ignore[no-redef, type-var]
+        """
+        A Generic pydantic model to coerce to np.ndarray's.
+        """
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, val: Any, field: ModelField) -> Optional[np.ndarray]:
+            return _coerce2NDArray(cls, val, field)
+
+
+def _coerce2NDArray(cls: Type, val: Any, field: ModelField) -> np.ndarray:
+    if field.sub_fields is not None:
+        dtype_field = field.sub_fields[0]
+        return np.asarray(val, dtype=dtype_field.type_)
```

Comment:
FYI this is not new. This is part of the `NDArray` custom type class which I moved from `utils._types` to here. The `_coerce2NDArray` used to be called `_validate` (see below). I renamed it for consistency with `coerce_2tensor` etc.
https://github.com/SeldonIO/alibi-detect/blob/1af9fc053ac224a599f16fd78e0d894de12f89f1/alibi_detect/utils/_types.py#L50-L55

I am actually going to tweak these to be `coerce_2_ndarray` and `coerce_2_tensor`.

Reply:
RE how the work. They work a bit differently because `_coerece2NDArray` is called by the `NDArray` pydantic [custom data type ](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types), which is used in `schemas` as follows:

```
sigma: Optional[NDArray[np.float32]] = None
```

whilst `coerce_2tensor` is itself a validator, so is used like:

```
    # Validators
    _validate_flavour = validator('flavour', allow_reuse=True, pre=False)(validate_framework)
    _coerce_sigma2tensor = validator('sigma', allow_reuse=True, pre=False)(coerce_2tensor)
```

Wrt to what is happening in `dtype_field = field.sub_fields[0]`, the `np.float32` given to `NDArray` in my example is inserted as the first element in `field.sub_fields[0]`. The `val` is the actual value of `sigma` passed to the validator. Hence we coerce `val` to `np.float32`.

Reply:
Final bit of info. I did want to do `coerce_2tensor` in the same way as we do `NDArray` (i.e. as a custom data type), however I couldn't figure out a way to get the `flavour` info in there properly. So went for the less elegant validator approach for this.